File: src/main/java/fr/raconteur/simpleskinswapper/changeskin/SelectedSkinStore.java
package fr.raconteur.simpleskinswapper.changeskin;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonObject;
import com.mojang.authlib.properties.Property;
import fr.raconteur.simpleskinswapper.SimpleSkinSwapper;
import fr.raconteur.simpleskinswapper.gui.SkinType;
import net.fabricmc.loader.api.FabricLoader;
import net.minecraft.resources.Identifier;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Optional;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class SelectedSkinStore {

    private static final Path FILE = FabricLoader.getInstance().getGameDir()
            .resolve("skins").resolve("selected.json");
    private static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();

    // Single-threaded so writes stay in submission order without blocking callers (e.g. the render thread).
    private static final ExecutorService SAVE_EXECUTOR = Executors.newSingleThreadExecutor(r -> {
        Thread t = new Thread(r, "SimpleSkinSwapper-SkinStoreSave");
        t.setDaemon(true);
        return t;
    });

    private static volatile @Nullable Property selectedProperty = null;
    private static volatile boolean loaded = false;

    // In-memory only — for pause menu preview
    private static volatile @Nullable Identifier previewTextureId = null;
    private static volatile @Nullable SkinType previewSkinType = null;

    public static Optional<Property> get() {
        ensureLoaded();
        return Optional.ofNullable(selectedProperty);
    }

    public static void set(Property property) {
        selectedProperty = property;
        loaded = true;
        SAVE_EXECUTOR.submit(() -> save(property));
    }

    public static void setPreview(@Nullable Identifier textureId, @Nullable SkinType skinType) {
        previewTextureId = textureId;
        previewSkinType = skinType;
    }

    public static @Nullable Identifier getPreviewTexture() {
        return previewTextureId;
    }

    public static @Nullable SkinType getPreviewSkinType() {
        return previewSkinType;
    }

    private static synchronized void ensureLoaded() {
        if (loaded) return;
        loaded = true;
        try {
            if (!Files.exists(FILE)) return;
            String content = Files.readString(FILE);
            JsonObject obj = GSON.fromJson(content, JsonObject.class);
            if (obj == null || !obj.has("value")) return;
            String value = obj.get("value").getAsString();
            String signature = obj.has("signature") && !obj.get("signature").isJsonNull()
                    ? obj.get("signature").getAsString() : null;
            selectedProperty = new Property("textures", value, signature);
            SimpleSkinSwapper.LOGGER.info("SelectedSkinStore: loaded selected skin.");
        } catch (Exception e) {
            SimpleSkinSwapper.LOGGER.warn("SelectedSkinStore: failed to load: {}", e.getMessage());
        }
    }

    private static void save(Property property) {
        try {
            Files.createDirectories(FILE.getParent());
            JsonObject obj = new JsonObject();
            obj.addProperty("value", property.value());
            if (property.hasSignature()) {
                obj.addProperty("signature", property.signature());
            } else {
                obj.add("signature", com.google.gson.JsonNull.INSTANCE);
            }
            Files.writeString(FILE, GSON.toJson(obj));
        } catch (IOException e) {
            SimpleSkinSwapper.LOGGER.warn("SelectedSkinStore: failed to save: {}", e.getMessage());
        }
    }
}

File: src/main/kotlin/fr/raconteur/simpleskinswapper/changeskin/SelectedSkinStore.kt
package fr.raconteur.simpleskinswapper.changeskin

import com.mojang.authlib.properties.Property
import fr.raconteur.simpleskinswapper.SimpleSkinSwapper
import fr.raconteur.simpleskinswapper.data.JsonFileStore
import fr.raconteur.simpleskinswapper.gui.SkinType
import kotlinx.serialization.Serializable
import net.fabricmc.loader.api.FabricLoader
import net.minecraft.resources.Identifier
import java.util.Optional
import java.util.concurrent.Executors

object SelectedSkinStore {

    /** On-disk shape of selected.json (Gson era wrote "signature" as explicit null too). */
    @Serializable
    private data class SelectedSkinDto(val value: String? = null, val signature: String? = null)

    private val store = JsonFileStore(
        fileLabel = "selected.json",
        path = {
            FabricLoader.getInstance().gameDir.resolve("skins").resolve("selected.json")
        },
        serializer = SelectedSkinDto.serializer(),
        fresh = { SelectedSkinDto() },
    )

    private val SAVE_EXECUTOR = Executors.newSingleThreadExecutor { r ->
        Thread(r, "SimpleSkinSwapper-SkinStoreSave").apply { isDaemon = true }
    }

    @Volatile
    private var selectedProperty: Property? = null

    @Volatile
    private var loaded = false

    @Volatile
    private var previewTextureId: Identifier? = null

    @Volatile
    private var previewSkinType: SkinType? = null

    @JvmStatic
    fun get(): Optional<Property> {
        ensureLoaded()
        return Optional.ofNullable(selectedProperty)
    }

    @JvmStatic
    fun set(property: Property) {
        selectedProperty = property
        loaded = true
        SAVE_EXECUTOR.submit { store.save(SelectedSkinDto(property.value(), property.signature())) }
    }

    @JvmStatic
    fun setPreview(textureId: Identifier?, skinType: SkinType?) {
        previewTextureId = textureId
        previewSkinType = skinType
    }

    @JvmStatic
    fun getPreviewTexture(): Identifier? = previewTextureId

    @JvmStatic
    fun getPreviewSkinType(): SkinType? = previewSkinType

    @Synchronized
    private fun ensureLoaded() {
        if (loaded) return
        loaded = true
        // A file without a "value" (or corrupt/missing) means "no stored selection".
        val dto = store.load()
        val value = dto.value ?: return
        selectedProperty = Property("textures", value, dto.signature)
        SimpleSkinSwapper.LOGGER.info("SelectedSkinStore: loaded selected skin.")
    }
}
